### PR TITLE
Deprecate implicit reads/writes for sync/single

### PR DESCRIPTION
**Pull request by mppf**
This PR issues deprecation warnings for implicit reads from and writes to
sync/single variables.

For issue #16193.

Details:
 * Renamed resolveNormalCallAdjustAssignType to
   resolveNormalCallAdjustAssign and updated it to rewrite `=` calls on
   sync/single in compiler-generated `=` to avoid the deprecation warning
 * Adjusted resolveInitVar to rewrite initialization of sync/single
   variables in compiler-generated `init`/`init=` to avoid the
   deprecation warning
 * Added a deprecation warning in `insertInitConversion` when a
   sync/single is converted to its value type (this code applies to
   returns from functions among other things)
 * Added a deprecation warning in `addArgCoercion`when a sync/single is
   passed to a function expecting its value type
 * Removes the `syncCounter` example from the spec since atomically
   updating a counter is better done with an `atomic`.
 * Removed the spec section describing how `write??`/`read??` can be
   implicitly invoked
 * Adjusted `initCopyAfterTransfer` and `chpl__transferArray` to avoid
   warnings for default-inited arrays of sync/single
 * Adds warnings in `chpl_statementLevelSymbol` and `_cond_test` when
   applied to sync/single (these handle cases like a `myvar;` and `if
   myvar`.
 * Adds warnings in `init=`, `=`, `<=>`, and `+=` etc with lhs
   sync/single and rhs sync/single and also to `chpl__initCopy` for these
   types.
 
These changes were needed in earlier versions of this PR but seemed to be
improvements:
 * Besides adding deprecations, this PR adds an implementation of
   `_desync` for `atomic` types which allows the future
   test/types/atomic/vass/initialize-atomics-1.chpl to pass. `_desync` is
   used for `Array = Scalar` and `Array += Scalar` type functions.
 * Adds zero-argument initializers to `localesSignal`,
   `chpl_LocalSpinlock`, `chpl_root_locale_accum` to avoid
   compiler-generated initializers with atomic arguments.

Future Work:
 * check description of sync variables in language specification and
   update it as needed
 * look at the documentation page generated by the syncsyngle.chpl primer
   and improve it as needed
 * figure out what `new` patterns compiler-generated initializers should
   support for sync/single/atomic fields. Should you be able to
   initialize from a sync/single actual? What about the numeric type
   stored in it? Today, copy initialization from a sync/single waits for
   the input to be "full" but it might make sense instead to result in an
   "empty" sync/single instead.
 * likewise, figure what behavior we want when to copy-initializeing
   record or array containing sync, single, or atomic elements.
 * after the 1.24 release, remove sync/single -> value type implicit
   conversions entirely - at least remove them in `canCoerceToCopyType`.
  * create a locking library so many uses of sync variables as a lock can
    be replaced with something easier to understand - #9881
 
Developed with / reviewed by both @bradcray and @mppf.

- [x] full local testing

### CODE REVIEW

**Comment by bradcray** [comment]
@mppf: Thanks _so_ much for helping out with this this week.  I've reviewed the vast majority of this, but:

- [ ] should have you walk me through one of the compiler changes (unless the debug test run I'm about to do clarifies its purpose to me)
- [ ] should take another read-through of the spec changes (could happen after feature freeze)
- [ ] am curious how you decided when to update the .goods to add deprecation warnings vs. update the test code to use the explicit read??/write?? calls (where I would've probably done more of the latter once I felt confident we had a deprecation test that locked in the former)

I made the following obvious cleanups on my local copy while reviewing:

```patch
diff --git a/compiler/AST/AggregateType.cpp b/compiler/AST/AggregateType.cpp
index 0a7c951801..337ff1fe48 100644
--- a/compiler/AST/AggregateType.cpp
+++ b/compiler/AST/AggregateType.cpp
@@ -2402,14 +2402,12 @@ void AggregateType::fieldToArg(FnSymbol*              fn,
 
             arg->defaultExpr = new BlockStmt(defPoint->init->copy());
             arg->typeExpr = new BlockStmt(defPoint->exprType->copy());
-            //arg->typeExpr = new BlockStmt(new CallExpr("_desync_warn", defPoint->exprType->copy()));
 
           } else {
             fieldToArgType(defPoint, arg);
 
             arg->defaultExpr = new BlockStmt(defPoint->init->copy());
             arg->typeExpr = new BlockStmt(defPoint->exprType->copy());
-            //arg->typeExpr = new BlockStmt(new CallExpr("_desync_warn", defPoint->exprType->copy()));
           }
         }
 
@@ -2427,7 +2425,6 @@ void AggregateType::fieldToArg(FnSymbol*              fn,
 
 void AggregateType::fieldToArgType(DefExpr* fieldDef, ArgSymbol* arg) {
   BlockStmt* exprType = new BlockStmt(fieldDef->exprType->copy(), BLOCK_TYPE);
-  //BlockStmt* exprType = new BlockStmt(new CallExpr("_desync_warn", fieldDef->exprType->copy()), BLOCK_TYPE);
 
   // If the type is simple, just set the argument's type directly.
   // Otherwise, give it the block we just created.
diff --git a/modules/internal/ChapelArray.chpl b/modules/internal/ChapelArray.chpl
index 5cd01932f4..9574952301 100644
--- a/modules/internal/ChapelArray.chpl
+++ b/modules/internal/ChapelArray.chpl
@@ -3831,7 +3831,7 @@ module ChapelArray {
     return false;
   }
 
-  // these functions avoid spurios warnings related to sync variable
+  // these functions avoid spurious warnings related to sync variable
   // deprecation, so once the deprecation has completed these can be removed.
   private proc isCopyableOrSyncSingle(type t) param {
     return isSyncType(t) || isSingleType(t) || isCopyableType(t);
@@ -4299,12 +4299,6 @@ module ChapelArray {
   }
 
   inline proc =(ref a: [], b) /* b is not an array nor a domain nor a tuple */ {
-    /*
-    if isSyncType(a.eltType) {
-      compilerWarning("Direct assignment to an array of 'sync' elements is deprecated; apply '.writeEF()' to set the elements");
-    } else if isSingleType(a.eltType) {
-      compilerWarning("Direct assignment to an array of 'single' elements is deprecated; apply '.writeEF()' to set the elements");
-    }*/
     chpl__transferArray(a, b);
   }
 
@@ -4374,30 +4368,6 @@ module ChapelArray {
     return x.valType;
   }
 
-  /*
-  proc _desync_warn(type t:_syncvar) type {
-    if chpl_warnUnstable then
-      compilerWarning("In an upcoming release, compiler-generated default initializers for objects containing 'sync' fields will change from taking a formal of type 'sync t' to simply 't'.  You can stabilize your code from such changes by adding an explicit default initializer now.");
-    return t;
-  }
-
-  proc _desync_warn(type t:_singlevar) type {
-    if chpl_warnUnstable then
-      compilerWarning("In an upcoming release, compiler-generated default initializers for objects containing 'single' fields will change from taking a formal of type 'single t' to simply 't'.  You can insultate your code from this change by adding an explicit default initializer now.");
-    return t;
-  }
-
-  proc _desync_warn(type t) type where isAtomicType(t) {
-    if chpl_warnUnstable then
-      compilerWarning("In an upcoming release, compiler-generated default initializers for objects containing 'atomic' fields will change from taking a formal of type 'atomic t' to simply 't'.  You can insulate your code from this change by adding an explicit default initializer now.");
-    return t;
-  }
-
-  proc _desync_warn(type t) type {
-    return t;
-  }
-  */
-
   proc _desync(type t) type where isAtomicType(t) {
     var x: t;
     return x.read().type;
```

**Comment by mppf** [comment]
>  should have you walk me through one of the compiler changes (unless the debug test run I'm about to do clarifies its purpose to me)

Sure.

>  should take another read-through of the spec changes (could happen after feature freeze)

AFAIK the only changes were focused on getting testing working:
* updating tests
* in at least one case - deleting an outdated section describing functionality we no longer support

>  am curious how you decided when to update the .goods to add deprecation warnings vs. update the test code to use the explicit read??/write?? calls (where I would've probably done more of the latter once I felt confident we had a deprecation test that locked in the former)

The tests that I updated to include the warnings had names like "test sync assign" and so I thought it better to leave their code as-is. Of course we can make a different decision in individual cases. I was mostly trying to update to explicit calls if I thought it did not go against the purpose of the test.

> I made the following obvious cleanups on my local copy while reviewing:

Yep, I left these in for you to look at - they are things you had added that I removed and I wanted to call them out.

**Comment by gbtitus** [comment]
I didn't actually "review" in any real sense, but I've read through the back-and-forth above to get a sense of how confident you both are about direction and done-ness and think it would be great to get this in for the release, so: 👍

File: test/functions/vass/type-and-param-function-arg-conversions.chpl [hunk 1]
```diff
@@ -15,12 +15,12 @@ proc typeFunUse(i:int)    type  return [1..i] int;
 
 var sv1, sv2, sv3: sync int;
 
-writeln(paramFun(sv1));
+writeln(paramFun(sv1.readXX()));
```
Comment (mppf) [inline test/functions/vass/type-and-param-function-arg-conversions.chpl:18]: @bradcray - I see that `readXX` is necessary here (vs `readFE`) but it seems to me that adding it makes the test no longer check what it was intended too. The thing is, I'm not sure how to do the check it wants to without user-defined implicit conversions or using the deprecated sync operations.

File: test/parallel/sync/diten/userLevelEndCount.chpl [hunk 2]
```diff
@@ -1,29 +1,30 @@
 class EC {
   var i: sync int(64) = 0;
   var b: sync bool = true;
+  // TODO: compiler-generated initializer
```
Comment (mppf) [inline test/parallel/sync/diten/userLevelEndCount.chpl:4]: @bradcray - should we remove this comment?

File: test/parallel/taskPar/sungeun/private.chpl [hunk 3]
```diff
@@ -24,22 +24,22 @@ class localePrivateData {
   type myStuff;
   // assumes maxTaskPar is the same across all locales
   const numTasks = if dataParTasksPerLocale==0 then here.maxTaskPar
-    else dataParTasksPerLocale;
+  else dataParTasksPerLocale;
   var slot: sync bool;
   var r = {0..#numTasks};
-  var temps: [r] myStuff;
+  var temps: [r] myStuff;  // TODO: initializer
```
Comment (mppf) [inline test/parallel/taskPar/sungeun/private.chpl:30]: @bradcray - here is another test TODO comment added. Not sure what the TODO would be or if we should to it in this PR.

File: test/reductions/diten/reduction_name_collision.chpl [hunk 4]
```diff
@@ -13,12 +13,12 @@ delete s;
 delete r;
 
 class R {
-  var count: sync int = 0;
+  var count: sync int = 0;  // TODO: initializer
```
Comment (mppf) [inline test/reductions/diten/reduction_name_collision.chpl:16]: @bradcray - here too

File: test/functions/vass/type-and-param-function-arg-conversions.chpl (hunk 1 above)
Comment (bradcray) [reply test/functions/vass/type-and-param-function-arg-conversions.chpl:18]: Huh, interesting...  Is it clear how/why this worked previously, with the compiler-injected .readFE()?  Is it just that we resolve the param function prior to inserting the .readFE() call?  (but then why would the type be interpreted as `int` rather than `sync int`).

Let's see what @vasslitvinov would like to do here.  In some respects, I consider this a positive outcome in that it seems to make `param` and `type` functions more symmetrical in this respect (where I read the original test as saying "Is this seeming asymmetry weird?  Should we change something?").  Left to my own devices, I'd make the param version more symmetrical with the type cases by filling the variables prior to the call and using .readFE() (but apparently didn't spend enough time thinking about this to do it previously).

I love the description on the PR that merged it:  https://github.com/chapel-lang/chapel/pull/249  :D

File: modules/internal/ChapelSyncvar.chpl [hunk 5]
```diff
@@ -803,9 +831,21 @@ module ChapelSyncvar {
...
   }
 
+  proc chpl__compilerGeneratedAssignSyncSingle(ref lhs : _singlevar(?),
+                                               ref rhs : _singlevar(?)) {
+    // TODO: Should this clone the value and the full/empty state instead?
+    lhs.writeEF(rhs.readFF());
+  }
+
+  proc chpl__compilerGeneratedCopySyncSingle(ref sv : _singlevar(?)) {
+    // TODO: this should probably clone the value and full/empty state instead
```
Comment (bradcray) [inline modules/internal/ChapelSyncvar.chpl:841]: @mppf With respect to this TODO and the one just above:  For singles, these effectively are just cloning the full/empty state, right?  Or do you mean something more technical by "clone"?

Not critical for this PR, but I'm also curious what your thoughts are about the post-deprecation status of these routines.  This is arguably where I got hung up on my effort...  it didn't seem right to use a sync/single to initialize a sync/single which is what led me to try to pipe `t` values through initializers for fields of type `sync t`, so I got bogged down in trying to make that work rather than preserving the status quo.

It looks to me like perhaps only 1 test other than your new deprecation tests actually calls into these routines at execution time, and only ~18 others have the compiler generate them, but then don't use them.

File: test/reductions/diten/reduction_name_collision.chpl (hunk 4 above)
Comment (bradcray) [reply test/reductions/diten/reduction_name_collision.chpl:16]: Yes to removing all three of these TODO notes, thanks.  I believe that each of these was indicating a case where my branch wasn't working, and where I could choose to work around it by introducing an explicit initializer, but that I didn't want to take that step since I thought it was a case where the user shouldn't need to change anything.  At some point, I kept re-diagnosing the same failure cases so started making notes like this to myself to capture the diagnosis, I believe.

File: modules/internal/ChapelSyncvar.chpl (hunk 5 above)
Comment (mppf) [reply modules/internal/ChapelSyncvar.chpl:841]: @bradcray - 

>  For singles, these effectively are just cloning the full/empty state, right? Or do you mean something more technical by "clone"?

I might be misunderstanding your comment, but

``` chapel
    var ret: sv.type = sv.readFF();
```

doesn't clone the full/empty state - rather, it insists that the full/empty state is full and then "clones" the value.
If we were cloning the full/empty state, I would expect that - if the input variable `sv` is empty - then the output variable will also be empty.

I'm not really sure where the future will take us here. It seems like we can copy-initialize sync/single variables but it's not really clear which meaning we should give it (clone-allowing-empty vs the current wait-until-full-then-clone). I noted this as Future Work in the PR description though - and I think it's worth noting that the issue is present both for compiler-generated initializers as well as arrays or other records containing sync/single. In both cases we want to run some sort of default copy operation but it's not so clear what that is for sync/single.

Anyway, I suspect that clone-allowing-empty is what we want and won't be surprising, but I am happy we are not trying to do it in this PR.

Comment (bradcray) [reply modules/internal/ChapelSyncvar.chpl:841]: > If we were cloning the full/empty state, I would expect that - if the input variable sv is empty - then the output variable will also be empty.

Ah, I see.  That makes sense, thanks.

File: test/classes/initializers/compilerGenerated/syncFieldDefaultInit.chpl [hunk 6]
```diff
@@ -1,5 +1,12 @@
 class Foo {
   var x$: sync int = 5;
+
+  // TODO: compiler should generate this (formal without sync)
+  /*
+  proc init(x: int = 5) {
+    this.x$ = x;
+  }
+  */
```
Comment (bradcray) [inline test/classes/initializers/compilerGenerated/syncFieldDefaultInit.chpl:9]: I think this arguably could be removed (and I think I should file an issue capturing the feature request).

Comment (mppf) [reply test/classes/initializers/compilerGenerated/syncFieldDefaultInit.chpl]: Alright, I've removed it. The issue would be a good one to link to from the PR description's future work section.

File: test/parallel/taskPar/figueroa/taskParallel.chpl [hunk 7]
```diff
@@ -225,15 +225,15 @@ proc determineWinner (player1, player2) {
...
   else if player1 == choices.paper then
     if player2 == choices.scissors then
       declareWinner("Player 2 wins!");
-    else declareWinner("Player 1 wins!");
+  else declareWinner("Player 1 wins!");
   else if player1 == choices.scissors then
     if player2 == choices.rock then
       declareWinner("Player 2 wins!");
-    else declareWinner("Player 1 wins!");
+  else declareWinner("Player 1 wins!");
```
Comment (bradcray) [inline test/parallel/taskPar/figueroa/taskParallel.chpl:236]: These three new indentations look "less correct" to me.  That said, I'm not deeply worried about this test's formatting.

Comment (bradcray) [reply test/parallel/taskPar/figueroa/taskParallel.chpl:236]: (Also: I see that the mistake was mine).

Comment (mppf) [reply test/parallel/taskPar/figueroa/taskParallel.chpl:236]: I can fix these indents.

File: test/release/examples/primers/syncsingle.chpl [hunk 8]
```diff
@@ -21,76 +21,19 @@ config const n = 7;
 var sy$: sync int=1;  // state = full, value = 1
 var si$: single bool; // state = empty, value = false
 
-// By convention, sync and single variable names end in a ``$`` to
-// add a visual cue indicating that reads and writes are possibly
-// expensive operations and can potentially block the task.
+// As a convention, sync and single variable names sometimes end in a
+// ``$`` to add a visual cue indicating that reads and writes are possibly
+// expensive operations that can potentially block the task.
```
Comment (bradcray) [inline test/release/examples/primers/syncsingle.chpl:26]: Not related to this PR, but I wonder whether we'll relax this convention in the near (or distant) future given that accidentally referring to a sync is less problematic (and all methods on variables can be potentially expensive).

File: test/trivial/deitz/statements/multi_decl4.chpl [hunk 9]
```diff
@@ -1,7 +1,19 @@
...
+     var x: sync int;
+     var y: sync int;
+*/
+
+// Works:
+//   var x: sync int, y: sync int;
+
+
+x.writeEF(1);
+y.writeEF(2);
```
Comment (bradcray) [inline test/trivial/deitz/statements/multi_decl4.chpl:16]: It looks to me like we could revert the changes to this test.  I suspect my original rewrite was to keep it working at some point that it wasn't, and am not seeing any reason for the change anymore?

**Comment by bradcray** [comment]
I just finished a full pass of the diffs, where I think the open questions are:
* what to do about the param/type proc test of Vass's
* whether we can/should revert `test/trivial/deitz/statements/multi_decl4.chpl`

I still skimmed over the changes to the docs (spec and primers), but still think I can do that post-merge

File: test/trivial/deitz/statements/multi_decl4.chpl (hunk 9 above)
Comment (mppf) [reply test/trivial/deitz/statements/multi_decl4.chpl]: I've reverted it, thanks.
